contributors/design-proposals/loadbalancer-proposal.md

```diff
@@ -0,0 +1,396 @@
+which means multiple LoadBalancerClaim can be bound to one LoadBalancer. For
+example, if we have a LoadBalancer with 3G bandwidth, we can bind 6 LoadBalancerClaim
+each request 500m bandwidth on it. In such case, we need to a 'scheduling' logic.
+
+Further, we can eventually introduce the 'request/limit model' for network resource
+to acheieve functionalities already implemented in compute resource, for example,
+qos and overcommit.
+
+##### Manually assign LoadBalancerClaim to a LodBalancer
+User can also manually assign LoadBalancerClaim to a LodBalancer, instead of letting
+loadbalancer-controller schedule for him. In such a case, resource request of all
+loadbalancerclaims may excess the loadbalancer's capacity. We validate request and
+capacity when loadbalancer-controller updating the loadbalancing rules: sort all
+Ingress "consume" a same LoadBalancers by creation time, if the sum request excess
+loadbalancer's capacity, avoid updating rules for the last few Ingress and send a
+event. Just like how we validate request at kubelet's side for Pods.
+
+
+## API
+
+### Network Resource
+
+```go
+const (
+	ResourceBandWidth ResourceName = "network-bandwidth"
+	ResourceIOPS      ResourceName = "network-iops"
+)
+```
+
+We can introduce more network resource in the future.
+
+### Loadbalancer API
+
+```go
+type LoadBalancer struct {
+	unversioned.TypeMeta `json:",inline"`
+	ObjectMeta           `json:"metadata,omitempty"`
+
+	// Spec defines a loadbalancer owned by the cluster
+	Spec LoadBalancerSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about loadbalancer.
+	Status LoadBalancerStatus `json:"status,omitempty"`
+}
+
+type LoadBalancerSpec struct {
+	// Resources represents the actual resources of the loadbalancer
+	Capacity ResourceList `json:"capacity"`
+	// Source represents the location and type of a loadbalancer to use.
+	LoadBalancerSource `json:",inline"`
+}
+
+type LoadBalancerSource struct {
+	GCELoadBalancer       *GCELoadBalancerSource       `json:"gceLoadBalancer,omitempty"`
+	AWSLoadBalancer       *AWSLoadBalancerSource       `json:"awsLoadBalancer,omitempty"`
+	BareMetalLoadBalancer *BareMetalLoadBalancerSource `json:"bareMetalLoadBalancer,omitempty"`
+	/*
+		more loadbalancer source
+	*/
+}
+
+type GCELoadBalancerSource struct {
+	// Unique name of the LoadBalancer resource. Used to identify the LoadBalancer in GCE
+	LBName string `json:"lbName"`
+}
+
+type AWSLoadBalancerSource struct {
+	// Unique name of the LoadBalancer resource. Used to identify the LoadBalancer in AWS
+	LBName string `json:"lbName"`
+}
+
+type BareMetalLoadBalancerSource struct {
+	Type        BareMetalLoadBalancerType `json:"type"`
+	ServiceName string                    `json:"serviceName"` //BareMetalLoadBalancer is actually a nginx or haproxy app deploymented in "kube-system" namespace
+}
+
+type BareMetalLoadBalancerType string
+
+const (
+	NginxLoadBalancer   BareMetalLoadBalancerType = "Nginx"
```

Comment:
See comment above about cloud providers. The feature should be load balancer type agnostic.